src/Transition.js

```diff
@@ -2,284 +2,274 @@ import React from 'react';
+    // This shouldn't be necessary, but there are weird race conditions with
+    // setState callbacks and unmounting in testing, so always make sure that
+    // we can cancel any pending setState callbacks after we unmount.
+    this.setState(nextState, this.setNextCallback(callback));
   }
 
-  render() {
-    let childProps = omit(this.props, Object.keys(Transition.propTypes).concat('children'));
+  setNextCallback(callback) {
+    let active = true;
 
-    let child = this.props.children;
-    let starting = this.state.needInitialRender;
-    let out = !this.state.in && !this.state.transitioning;
+    this.nextCallback = (event) => {
+      if (active) {
+        active = false;
+        this.nextCallback = null;
 
-    if ( !child || (this.props.unmountOnExit && out && !starting) ){
-      return null;
-    }
+        callback(event);
+      }
+    };
 
-    let classes = '';
+    this.nextCallback.cancel = () => {
+      active = false;
+    };
 
-    // using `classnames()` here causes a subtle bug,
-    // hence the verbose if/else if sequence.
-    if (this.state.in && !this.state.transitioning) {
-      classes = this.props.enteredClassName;
-    }
+    return this.nextCallback;
+  }
+
+  onTransitionEnd(node, handler) {
+    this.setNextCallback(handler);
 
-    else if (this.state.in && this.state.transitioning) {
-      classes = this.props.enteringClassName;
+    if (node) {
+      TransitionEvents.addEndEventListener(node, this.nextCallback);
+      setTimeout(this.nextCallback, this.props.duration);
+    } else {
+      setTimeout(this.nextCallback, 0);
     }
+  }
 
-    else if (!this.state.in && !this.state.transitioning) {
-      classes = this.props.exitedClassName;
+  render() {
+    const status = this.state.status;
+    if (status === UNMOUNTED) {
+      return null;
     }
 
-    else if (!this.state.in && this.state.transitioning) {
-      classes = this.props.exitingClassName;
+    const {children, className, style, ...childProps} = this.props;
+    Object.keys(Transition.propTypes).forEach(key => delete childProps[key]);
+
+    let transitionClassName;
+    if (status === EXITED) {
+      transitionClassName = this.props.exitedClassName;
+    } else if (status === ENTERING) {
+      transitionClassName = this.props.enteringClassName;
+    } else if (status === ENTERED) {
+      transitionClassName = this.props.enteredClassName;
+    } else if (status === EXITING) {
+      transitionClassName = this.props.exitingClassName;
     }
 
-    return React.cloneElement(child, {
-      ...childProps,
-      className: classnames(
-        child.props.className,
-        this.props.className,
-        classes)
-    });
+    const child = React.Children.only(children);
+    return React.cloneElement(
+      child,
+      {
+        ...childProps,
+        className: classnames(
+          child.props.className,
+          className,
+          transitionClassName
+        ),
+        style: {...child.props.style, ...style}
```

Comment:
why this? My thought was that if someone wanted to style the child they can just add it there, instead of introducing an additional input vector. I don't have any strong opinion here, just wondering